File: CPP_Joc_Windows_Android/SH_Client_Win_Cpp_Cmake/App/src/base/crypto/algorithm/Salsa20Algorithm.cpp
#include "Salsa20Algorithm.h"

using namespace base::crypto;

Salsa20Algorithm::Salsa20Algorithm(const std::uint8_t* key, const std::uint8_t* iv)
    : rawAlgorithm()
{
    rawAlgorithm.setKey(key);
    rawAlgorithm.setIv(iv);
}

void Salsa20Algorithm::processBytes(const std::uint8_t* src, std::uint8_t* dest, int count) {
    rawAlgorithm.processBytes(src, dest, count);
}

Salsa20Algorithm::~Salsa20Algorithm() {
    //void
}

File: NodeJs_TypeScript_JavaScript_Server/Project_NodeJs_VSCode/src/base/crypto/algorithm/Salsa20Algorithm.ts
import { IAlgorithm } from "./IAlgorithm";
import { Salsa20 } from "../rawalgorithm/Salsa20";

export class Salsa20Algorithm implements IAlgorithm {

    // Values are the ones from the raw algorithm. They are used to validate the key and iv sizes;
    // base::crypto::Salsa20::size_t::KEY_SIZE;
    public static readonly KEY_SIZE = 32;
    // base::crypto::Salsa20::size_t::IV_SIZE;
    public static readonly IV_SIZE = 8;
    
    private rawAlgorithm: Salsa20;

    public constructor(key: Uint8Array, iv: Uint8Array) {
        this.rawAlgorithm = new Salsa20();
        this.rawAlgorithm.setKeyIv(key, iv);
    }
    
    public processBytes(inBuf: ArrayBuffer, inBytesOffset: number, 
        outBuf: ArrayBuffer, outBytesOffset: number, 
        bytesCount: number): void
    {
        this.rawAlgorithm.processBytes(
            inBuf, inBytesOffset, 
            outBuf, outBytesOffset, 
            bytesCount
        );
    }

}
